File: lib-universalAdapter/src/main/java/com/llj/adapter/ListBasedAdapter.java
package com.llj.adapter;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.view.ViewGroup;
import com.llj.adapter.model.ISelect;
import com.llj.adapter.observable.ObservableList;
import com.llj.adapter.observable.ObservableListWrapper;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;

/**
 * PROJECT:UniversalAdapter
 *
 * DESCRIBE:
 *
 * ObservableList：实现该接口以封装对集合的操作
 *
 * @Created by llj on 2017/2/10.
 */

public abstract class ListBasedAdapter<Item, Holder extends ViewHolder> extends
    UniversalAdapter<Item, Holder> implements ObservableList<Item> {

  private List<Item> mList;
  private int        mClickPosition;

  public ListBasedAdapter() {
    this(null);
  }

  public ListBasedAdapter(List<Item> list) {
    setItemsList(list);
  }

  public ListBasedAdapter(ObservableListWrapper<Item> list) {
    setItemsList(list);
  }


  public int getClickPosition() {
    return mClickPosition;
  }

  public void setClickPosition(int clickPosition) {
    mClickPosition = clickPosition;
  }

  public Item getClickItem() {
    if (size() > mClickPosition) {
      return get(mClickPosition);
    }
    return null;
  }

  public void setItemsList(ObservableList<Item> list) {
    if (list != null) {
      list.getListObserver().addListener(observableListener);
    }
    setItemsList((List<Item>) list);
  }

  protected void unbindList() {
    if (mList instanceof ObservableList<?>) {
      ((ObservableList<Item>) mList).getListObserver().removeListener(observableListener);
    }
  }

  public void setItemsList(List<Item> list) {
    unbindList();
    if (list == null) {
      list = new LinkedList<>();
    }
    mList = list;
    notifyDataSetChanged();
  }


  public List<Item> getList() {
    return mList;
  }

  ///////////////////////////////////////////////////////////////////////////
  //
  ///////////////////////////////////////////////////////////////////////////
  @Override
  public int size() {
    return mList.size();
  }

  @Override
  public int getCount() {
    return mList.size();
  }

  @Override
  public boolean isEmpty() {
    return mList.isEmpty();
  }


  ///////////////////////////////////////////////////////////////////////////
  // 增删改对应刷新操作
  ///////////////////////////////////////////////////////////////////////////
  @Override
  public void add(int index, Item object) {
    mList.add(index, object);
    onItemRangeInserted(index, 1);
  }

  @Override
  public boolean add(Item object) {
    int location = mList.size();
    final boolean result = mList.add(object);
    onItemRangeInserted(location, 1);
    return result;
  }

  private boolean isCollectionEmpty(Collection<? extends Item> list) {
    return list == null || list.isEmpty();
  }

  @Override
  public boolean addAll(@NonNull Collection<? extends Item> collection) {
    if (isCollectionEmpty(collection)) {
      return false;
    }
    int location = mList.size();
    if (mList.addAll(collection)) {
      onItemRangeInserted(location, collection.size());
      return true;
    }
    return false;
  }

  @Override
  public boolean addAll(int index, @NonNull Collection<? extends Item> collection) {
    if (isCollectionEmpty(collection)) {
      return false;
    }
    if (mList.addAll(index, collection)) {
      onItemRangeInserted(index, collection.size());
      return true;
    }
    return false;
  }

  @Override
  public Item remove(int index) {
    Item result = mList.remove(index);
    onItemRangeRemoved(index, 1);
    return result;
  }

  @Override
  public boolean remove(@NonNull Object object) {
    int location = mList.indexOf(object);
    if (location >= 0) {
      mList.remove(location);
      onItemRangeRemoved(location, 1);
      return true;
    }
    return false;
  }

  @Override
  public boolean removeAll(@NonNull Collection<?> collection) {
    boolean result = mList.removeAll(collection);
    if (result) {
      onGenericChange();
    }
    return result;
  }

  @Override
  public boolean retainAll(@NonNull Collection<?> collection) {
    boolean result = mList.retainAll(collection);
    if (result) {
      onGenericChange();
    }
    return result;
  }

  @Override
  public void clear() {
    int count = size();
    mList.clear();
    onItemRangeRemoved(0, count);
  }


  @Override
  public Item set(int index, Item object) {
    Item result = mList.set(index, object);
    if (!result.equals(object)) {
      onItemRangeChanged(index, 1);
    }
    return result;
  }

  @Nullable
  @Override
  public Item get(int position) {
    return mList.get(position);
  }

  ///////////////////////////////////////////////////////////////////////////
  //
  ///////////////////////////////////////////////////////////////////////////
  @Override
  public boolean contains(Object object) {
    return mList.contains(object);
  }

  @Override
  public boolean containsAll(@NonNull Collection<?> collection) {
    return mList.containsAll(collection);
  }

  @NonNull
  @Override
  public Object[] toArray() {
    if (mList == null) {
      return new ArrayList<>().toArray();
    }
    return mList.toArray();
  }

  @NonNull
  @Override
  public <T> T[] toArray(@NonNull T[] array) {
    return mList.toArray(array);
  }


  @Override
  public int indexOf(Object object) {
    return mList.indexOf(object);
  }

  @Override
  public int lastIndexOf(Object object) {
    return mList.lastIndexOf(object);
  }

  @NonNull
  @Override
  public Iterator<Item> iterator() {
    return mList.iterator();
  }

  @NonNull
  @Override
  public ListIterator<Item> listIterator() {
    return mList.listIterator();
  }

  @NonNull
  @Override
  public ListIterator<Item> listIterator(int index) {
    return mList.listIterator();
  }

  @NonNull
  @Override
  public List<Item> subList(int start, int end) {
    return mList.subList(start, end);
  }


  ///////////////////////////////////////////////////////////////////////////
  //
  ///////////////////////////////////////////////////////////////////////////
  @Override
  protected Holder onCreateViewHolder(@NonNull ViewGroup parent, int itemType) {
    return null;
  }


  @Override
  public void notifyDataSetChanged() {
    onGenericChange();
  }


  public Item getLast() {
    if (getList() == null || getList().isEmpty()) {
      return null;
    }
    return getList().get(size() - 1);
  }

  public void removeLast() {
    if (getList() == null || getList().isEmpty()) {
      return;
    }
    getList().remove(size() - 1);
  }


  //获取选中项
  public ISelect getSelectItem() {
    List<Item> itemsList = getList();
    if (isCollectionEmpty(itemsList)) {
      return null;
    }
    for (Item item : itemsList) {
      if (item == null) {
        continue;
      }
      if (item instanceof ISelect && ((ISelect) item).isSelect()) {
        return ((ISelect) item);
      }
    }
    return null;
  }
}

File: lib-universalAdapter/src/main/java/com/llj/adapter/ListBasedAdapter.kt
package com.llj.adapter

import com.llj.adapter.model.ISelect
import com.llj.adapter.observable.ListObserver
import com.llj.adapter.observable.ObservableList

/**
 * PROJECT:UniversalAdapter
 *
 * DESCRIBE:
 *
 * ObservableList：实现该接口以封装对集合的操作
 *
 * @Created by llj on 2017/2/10.
 */
abstract class ListBasedAdapter<Item, Holder : XViewHolder> : UniversalAdapter<Item?, Holder>, ObservableList<Item?> {

  private lateinit var mList: MutableList<Item?>

  var clickPosition = 0

  constructor() : super() {
    setItemsList(null)
  }

  constructor(list: MutableList<Item?>) : super() {
    setItemsList(list)
  }


  val clickItem: Item?
    get() = if (size > clickPosition) {
      get(clickPosition)
    } else null


  fun setItemsList(list: ObservableList<Item?>) {
    list.getListObserver().addListener(observableListener)
    setItemsList(list as MutableList<Item?>)
  }


  fun unbindList() {
    if (mList is ObservableList<Item?>) {
      (mList as ObservableList<Item?>).getListObserver().removeListener(observableListener)
    }
  }

  fun setItemsList(list: MutableList<Item?>?) {
    var listNew = list
//    unbindList()
    if (listNew == null) {
      listNew = ArrayList()
    }
    mList = listNew
    notifyDataSetChanged()
  }

  fun getList(): MutableList<Item?> {
    return mList
  }

  override fun getListObserver(): ListObserver<Any?> {
    return mListObserver
  }

  ///////////////////////////////////////////////////////////////////////////
  //
  ///////////////////////////////////////////////////////////////////////////

  //List
  override val size: Int
    get() = mList.size


  override fun getCount(): Int {
    return mList.size
  }

  //List
  override fun isEmpty(): Boolean {
    return mList.isEmpty()
  }

  private fun isCollectionEmpty(list: Collection<Item?>?): Boolean {
    return list == null || list.isEmpty()
  }

  //region  ObservableList<Item?>
  override fun add(index: Int, element: Item?) {
    mList.add(index, element)
    onItemRangeInserted(index, 1)
  }

  override fun add(element: Item?): Boolean {
    val location = mList.size
    val result = mList.add(element)
    onItemRangeInserted(location, 1)
    return result
  }

  override fun addAll(elements: Collection<Item?>): Boolean {
    if (isCollectionEmpty(elements)) {
      return false
    }
    val location = mList.size
    if (mList.addAll(elements)) {
      onItemRangeInserted(location, elements.size)
      return true
    }
    return false
  }

  override fun addAll(index: Int, elements: Collection<Item?>): Boolean {
    if (isCollectionEmpty(elements)) {
      return false
    }
    if (mList.addAll(index, elements)) {
      onItemRangeInserted(index, elements.size)
      return true
    }
    return false
  }

  override fun removeAt(index: Int): Item? {
    val result = mList.removeAt(index)
    onItemRangeRemoved(index, 1)
    return result
  }

  override fun remove(element: Item?): Boolean {
    val location = mList.indexOf(element)
    if (location >= 0) {
      removeAt(location)
      return true
    }
    return false
  }

  override fun removeAll(elements: Collection<Item?>): Boolean {
    val result = mList.removeAll(elements)
    if (result) {
      onGenericChange()
    }
    return result
  }

  override fun retainAll(elements: Collection<Item?>): Boolean {
    val result = mList.retainAll(elements)
    if (result) {
      onGenericChange()
    }
    return result
  }

  override fun clear() {
    val count = size
    mList.clear()
    onItemRangeRemoved(0, count)
  }

  override fun set(index: Int, element: Item?): Item? {
    val result = mList.set(index, element)
    if (result != element) {
      onItemRangeChanged(index, 1)
    }
    return result
  }
  //endregion


  //region List<Item?>
  override fun get(index: Int): Item? {
    if (index >= mList.size) {
      return null
    }
    return mList[index]
  }

  override operator fun contains(element: Item?): Boolean {
    return mList.contains(element)
  }

  override fun containsAll(elements: Collection<Item?>): Boolean {
    return mList.containsAll(elements)
  }

  override fun indexOf(element: Item?): Int {
    return mList.indexOf(element)
  }

  override fun lastIndexOf(element: Item?): Int {
    return mList.lastIndexOf(element)
  }

  override fun iterator(): MutableIterator<Item?> {
    return mList.iterator()
  }

  override fun listIterator(): MutableListIterator<Item?> {
    return mList.listIterator()
  }

  override fun listIterator(index: Int): MutableListIterator<Item?> {
    return mList.listIterator()
  }

  override fun subList(fromIndex: Int, toIndex: Int): MutableList<Item?> {
    return mList.subList(fromIndex, toIndex)
  }
  //endregion

  ///////////////////////////////////////////////////////////////////////////
  //
  ///////////////////////////////////////////////////////////////////////////

  override fun notifyDataSetChanged() {
    onGenericChange()
  }

  //获取最后一个item
  fun getLast(): Item? {
    if (getList().isEmpty()) {
      return null
    }
    return getList()[size - 1]

  }

  //移除最后一个item
  fun removeLast() {
    if (getList().isEmpty()) {
      return
    }
    getList().removeAt(size - 1)
  }

  //获取选中项
  val selectItem: ISelect?
    get() {
      val itemsList = getList()
      if (isCollectionEmpty(itemsList)) {
        return null
      }
      for (item in itemsList) {
        if (item == null) {
          continue
        }
        if (item is ISelect && (item as ISelect).isSelect) {
          return item
        }
      }
      return null
    }

}